content/docs/tasks/security/vault-ca/index.md

```diff
@@ -154,3 +158,123 @@ between workloads using the certificates the Vault CA issued.
+based on the Vault Kubernetes auth method.
+**The instructions here are for illustrative purposes only. Please consult with security experts
+on the security configuration of your Vault servers.** The instructions are based on the
+post [1](https://evalle.xyz/posts/integration-kubernetes-with-vault-auth/) and
+[2](https://coreos.com/tectonic/docs/latest/vault-operator/user/kubernetes-auth-backend.html).
+
+1.  Create a Kubernetes cluster to host an example basic Vault server.
+    In the Kubernetes cluster created, install, initialize, unseal, and login Vault.
+    Examples of install, initialize, unseal, and login Vault can be found in the post
+    [1](https://evalle.xyz/posts/integration-kubernetes-with-vault-auth/).
+    The example Vault server used in this guide is of 0.10.3 version.
+
+1.  Follow the post [2](https://coreos.com/tectonic/docs/latest/vault-operator/user/kubernetes-auth-backend.html)
+    to set up a Kubernetes service account for Vault token review.
+
+    {{< text bash >}}
+    $ kubectl create serviceaccount vault-tokenreview
+    $ kubectl apply -f - <<EOF
+    apiVersion: rbac.authorization.k8s.io/v1beta1
+    kind: ClusterRoleBinding
+    metadata:
+      name: vault-tokenreview-binding
+      namespace: default
+    roleRef:
+      apiGroup: rbac.authorization.k8s.io
+      kind: ClusterRole
+      name: system:auth-delegator
+    subjects:
+    - kind: ServiceAccount
+      name: vault-tokenreview
+      namespace: default
+    EOF
+    $ SECRET_NAME=$(kubectl get serviceaccount vault-tokenreview -o jsonpath='{.secrets[0].name}')
+    $ TR_ACCOUNT_TOKEN=$(kubectl get secret ${SECRET_NAME} -o jsonpath='{.data.token}' | base64 --decode)
+    {{< /text >}}
+
+1.  Follow the post [2](https://coreos.com/tectonic/docs/latest/vault-operator/user/kubernetes-auth-backend.html)
+    to enable and configure the Vault Kubernetes Auth method. The parameters `kubernetes_host`
+    and `kubernetes_ca_cert` are described in the
+    [Vault API document](https://www.vaultproject.io/api/auth/kubernetes/index.html#kubernetes_host).
+
+    {{< text bash >}}
+    $ vault auth enable kubernetes
+    $ vault write auth/kubernetes/config kubernetes_host=<your-kubernetes-host> \
+        kubernetes_ca_cert=<your-kubernetes-ca-cert> token_reviewer_jwt=$TR_ACCOUNT_TOKEN
+    {{< /text >}}
+
+1.  Follow the post [2](https://coreos.com/tectonic/docs/latest/vault-operator/user/kubernetes-auth-backend.html)
+    to create a Vault policy and a role bound to the `default` service account.
+
+    {{< text bash >}}
+    $ cat <<EOF > ./policy.hcl
+    {
+        "name": "istio-cert",
+        "path": {
+            "istio_ca/sign/istio-pki-role": {
+                "capabilities": ["update", "read"]
+            }
+        }
+    }
+    EOF
+    $ vault write sys/policy/istio-cert policy=@./policy.hcl
+    $ vault write auth/kubernetes/role/istio-cert \
+        bound_service_account_names=default \
```

Comment:
Why are we using the `default` service account here?

Looking at the helm template for the `nodeagent` daemonset, it appears to be using a service account named `istio-nodeagent-service-account`.

https://github.com/istio/istio/blob/ed099bc2fefd3dbef9319634ab507738608663d4/install/kubernetes/helm/istio/charts/nodeagent/templates/daemonset.yaml#L27

So shouldn't we use that as the service account name rather than `default`?  Then you shouldn't need to do the patching of secrets up above.

If you change that, we should also change `bound_service_account_namespaces` to `istio-system`

Reply:
The k8s service account used for Vault authentication is from a workload pod (e.g., htttpbin, sleep), so istio-nodeagent-service-account is not used.

Reply:
Does that mean that to use Vault as a CA, I need to enable every list every single service account that will be used for workloads running in my cluster? That doesn't seem scalable at all.

My understanding is the workload makes a request to the node-agent, and then the node-agent is the one authenticating with vault to sign the CSR. So why isn't the node-agent using its service account to do that?

Reply:
The current design is that Vault authenticates the k8s service account from a workload. IMO, the approach that node-agent uses its service account to authenticate with Vault may have a few issues:
1. Different workloads on the node share the same node agent service account, while they may run with their individual service accounts and need certificates corresponding to their individual service accounts.
2. There may be multiple nodes. If different node agents have different service accounts, these service accounts still need to be configured on Vault. If different node agents share the same service account, then the security of the system hinges on this single node agent service account, which may not be very secure.

Different designs come with different tradeoffs, some approaches are easier to configure but less secure while other approaches take more efforts to configure but is more secure.

Reply:
So the workload passes its service account token to node-agent, and node-agent passes that along to Vault? Isn't that a separate security concern, where you've now allowed node-agent to authenticate to the Kubernetes API with different permissions?

I recognize I'm challenging a design decision in a documentation review, which probably isn't the right approach.  I'll open up a separate issue to discuss this.

Is this requirement (that the Vault role needs to include a list of all whitelisted service accounts or namespaces) documented somewhere? IMHO, that needs to be clearly spelled out.

Reply:
node-agent does not authenticate to the Kubernetes API and does not authenticate to Vault neither; instead, node-agent passes the k8s service account from a workload to Vault and Vault does the authentication. 
It is up to the Vault owners to decide their authentication and certificate issuance policies. There is no requirement that Vault must include a list of whitelisted service accounts or namespaces; Vault owners can enforce their own authentication and authorization policies based on the k8s service account and the CSR. The instructions here only configures a basic Vault server for illustration purpose based on Vault k8s auth method.

Reply:
> There is no requirement that Vault must include a list of whitelisted service accounts or namespaces

To configure the k8s auth method in Vault, you need to create a role. According to the [Vault docs for creating a role](https://www.vaultproject.io/api/auth/kubernetes/index.html#create-role), you need to specify one or both of:

> `bound_service_account_names (array: <required>)` - List of service account names able to access this role. If set to "*" all names are allowed, both this and bound_service_account_namespaces can not be "*".
> `bound_service_account_namespaces (array: <required>)` - List of namespaces allowed to access this role. If set to "*" all namespaces are allowed, both this and bound_service_account_names can not be set to "*". 

Since they both can't be `*`, it is necessary to have a list of either service account names or namespaces. That's what I mean by requiring a whitelist.

If nodeagent was passing in its own ServiceAccount credentials, I could just use the nodeagent's service account in `bound_service_account_names`. But since it's the workload's tokens being passed through, that won't work.

> The instructions here only configures a basic Vault server for illustration purpose based on Vault k8s auth method.

But as far as I can tell, this is the _only_ documentation there is on the Istio/Vault integration. Does NodeAgent support other methods for authenticating with Vault, other than the k8s auth method? If so, is that documented anywhere?

Reply:
Node Agent does not mandate the authentication method on a Vault endpoint and the authentication methods used by Vault endpoints are transparent to Node Agent. Node Agent passes the k8s service account from a workload and the CSR to a Vault endpoint; the Vault owner decides the authentication and authorization methods, as well as certificate issurance policies. 
For example, one Vault owner may use Vault k8s auth method, whereas another Vault owner may use a proprietary auth method (e.g., through a proprietary Vault auth plugin) that for the k8s service account A, a certificate of identity B is issued.